File: A2Z_striver/1 Basics/1.2 Patterns/pattern_7/pattern_7.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print a Center-Aligned Star Pyramid.
// Example Input: n = 3
// Example Output:
//   *
//  ***
// *****
// Goal: O(N^2) Time | O(1) Space

class Solution
{
public:
    // ~ Approach: Multiple Inner Loops
    // Analysis: Time O(N^2) | Space O(1)
    void pattern7(int n)
    {
        for (int i = 0; i < n; i++)
        {

            // - Step 1: Print Leading Spaces
            // - Logic: As row (i) increases, spaces decrease.
            for (int j = 0; j < n - i - 1; j++)
            {
                cout << " ";
            }

            // - Step 2: Print Stars
            // - Logic: Odd numbers (1, 3, 5...) -> Formula 2*i + 1
            for (int j = 0; j < 2 * i + 1; j++)
            {
                cout << "*";
            }

            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 5;

    cout << "Test Case (n=" << n << "):\n";
    sol.pattern7(n);

    return 0;
}

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_7/pattern_7.py
import sys


class Solution:
    """
    Problem: Print a Center-Aligned Star Pyramid.
    Example Input: n = 3
    Example Output:
      *
     ***
    *****
    
    Constraints: 1 <= n <= 100
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Nested Loops (Algorithmic Standard)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern7_nested(self, n):
        for i in range(n):
            # // - Step 1: Print Leading Spaces
            # //   Formula: n - i - 1
            for j in range(n - i - 1):
                print(" ", end="")

            # // - Step 2: Print Stars
            # //   Formula: 2*i + 1 (Odd numbers: 1, 3, 5...)
            for j in range(2 * i + 1):
                print("*", end="")
            
            # // - Step 3 (Optional): Trailing Spaces
            # //   Some OAs require symmetry. If so, copy the first loop here.
            
            print()

    # // ~ Approach 2: Pythonic One-Liner
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern7_pythonic(self, n):
        for i in range(n):
            # // . Calculate spaces and stars using string math
            spaces = " " * (n - i - 1)
            stars = "*" * (2 * i + 1)
            print(spaces + stars)

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    N = 5
    
    print(f"--- Approach 1: Nested Loops (N={N}) ---")
    solver.pattern7_nested(N)
    
    print(f"\n--- Approach 2: Pythonic One-Liner (N={N}) ---")
    solver.pattern7_pythonic(N)